File: python/sort/bubble_sort.py
def bubble_sort_in_place(elements: list):
    """
    Sorts the input list in ascending order using the Bubble Sort algorithm.
    Bubble Sort makes n-1 passes, where in each pass, it swaps all the
    elements from 0 to k (where k is n - the pass number), moving the biggest to the end.
    It has a worst-case time complexity of O(n^2).

    :param elements: List of elements to test_sort
    :raises TypeError: If `elements` is not a list.
    """
    # Ensure the input parameter is a list
    if not isinstance(elements, list):
        raise TypeError("Input must be a list.")

    if len(elements) <= 1:
        return  # A list with one or zero elements is always sorted.

    last_index = len(elements) - 1

    # Iterate over the list n-1 times
    for k in range(last_index, 0, -1):
        swapped = False
        for index in range(k):
            # If adjacent elements are in the wrong order, swap them
            if elements[index] > elements[index + 1]:
                elements[index], elements[index + 1] = elements[index + 1], elements[index]
                swapped = True
        if not swapped:
            return


def bubble_sort(elements: list) -> list:
    """
    Returns a sorted copy of the input list using the Bubble Sort algorithm.
    Bubble Sort makes n-1 passes, where in each pass, it swaps all the
    elements from 0 to k (where k is n - the pass number), moving the biggest to the end.
    It has a worst-case time complexity of O(n^2).

    :param elements: List of elements to test_sort.
    :return: A sorted copy of the input list.
    """
    # Make a copy of the input list
    elements_copy = elements.copy()

    if len(elements_copy) <= 1:
        return elements_copy  # Early return if the list only has one or zero elements

    # Sort the copy using the bubble test_sort algorithm
    bubble_sort_in_place(elements_copy)

    return elements_copy

File: rust/src/sort/bubble_sort.rs
/// Sorts the input slice in ascending order using the Bubble Sort algorithm.
///
/// Bubble Sort makes n-1 passes, where in each pass, it swaps all the
/// elements from 0 to k (where k is n - the pass number), moving the biggest to the end.
/// It has a worst-case time complexity of O(n^2).
///
/// # Arguments
///
/// * `elements` - The slice of elements to sort.
///
/// # Example
///
/// ```
/// use rust::sort::bubble_sort::bubble_sort;
/// let mut v = vec![3, 2, 1];
/// bubble_sort(&mut v);
/// assert_eq!(v, vec![1, 2, 3]);
/// ```
pub fn bubble_sort<T: PartialOrd>(elements: &mut [T]) {
    // A slice with one or zero elements is always sorted.
    if elements.len() <= 1 {
        return;
    }
    let last_index = elements.len() - 1;

    // Iterate over the slice n-1 times
    for k in (0..=last_index).rev() {
        let mut swapped = false;
        for index in 0..k {
            if elements[index] > elements[index + 1] {
                elements.swap(index, index + 1);
                swapped = true;
            }
        }
        if !swapped {
            break;
        }
    }
}

test_sort_function!(bubble_sort);